File: examples/zig/src/raylib-wasm.zig
pub const Color = packed struct {
    r: u8,
    g: u8,
    b: u8,
    a: u8 = 255,
};


// Some Basic Colors
// NOTE: Custom raylib color palette for amazing visuals on WHITE background
pub const LIGHTGRAY = Color{.r=200, .g=200, .b=200 };  // Light Gray
pub const GRAY      = Color{.r=130, .g=130, .b=130 };  // Gray
pub const DARKGRAY  = Color{.r=80, .g=80, .b=80 };     // Dark Gray
pub const YELLOW    = Color{.r=253, .g=249, .b=0 };    // Yellow
pub const GOLD      = Color{.r=255, .g=203, .b=0 };    // Gold
pub const ORANGE    = Color{.r=255, .g=161, .b=0 };    // Orange
pub const PINK      = Color{.r=255, .g=109, .b=194 };  // Pink
pub const RED       = Color{.r=230, .g=41, .b=55 };    // Red
pub const MAROON    = Color{.r=190, .g=33, .b=55 };    // Maroon
pub const GREEN     = Color{.r=0, .g=228, .b=48 };     // Green
pub const LIME      = Color{.r=0, .g=158, .b=47 };     // Lime
pub const DARKGREEN = Color{.r=0, .g=117, .b=44 };     // Dark Green
pub const SKYBLUE   = Color{.r=102, .g=191, .b=255 }; // Sky Blue
pub const BLUE      = Color{.r=0, .g=121, .b=241 };    // Blue
pub const DARKBLUE  = Color{.r=0, .g=82, .b=172 };     // Dark Blue
pub const PURPLE    = Color{.r=200, .g=122, .b=255 };  // Purple
pub const VIOLET    = Color{.r=135, .g=60, .b=190 };   // Violet
pub const DARKPURPLE= Color{.r=112, .g=31, .b=126 };   // Dark Purple
pub const BEIGE     = Color{.r=211, .g=176, .b=131 };  // Beige
pub const BROWN     = Color{.r=127, .g=106, .b=79 };   // Brown
pub const DARKBROWN = Color{.r=76, .g=63, .b=47 };     // Dark Brown
pub const WHITE     = Color{.r=255, .g=255, .b=255 }; // White
pub const BLACK     = Color{.r=0, .g=0, .b=0 };        // Black
pub const BLANK     = Color{.r=0, .g=0, .b=0, .a=0 };          // Blank (Transparent)
pub const MAGENTA   = Color{.r=255, .g=0, .b=255 };   // Magenta
pub const RAYWHITE  = Color{.r=245, .g=245, .b=245 };  // My own White (raylib logo)


// Trace log level
// NOTE: Organized by priority level
// pub enum TraceLogLevel {
//     LOG_ALL = 0,        // Display all logs
//     LOG_TRACE,          // Trace logging, intended for internal use only
//     LOG_DEBUG,          // Debug logging, used for internal debugging, it should be disabled on release builds
//     LOG_INFO,           // Info logging, used for program execution info
//     LOG_WARNING,        // Warning logging, used on recoverable failures
//     LOG_ERROR,          // Error logging, used on unrecoverable failures
//     LOG_FATAL,          // Fatal logging, used to abort program: exit(EXIT_FAILURE)
//     LOG_NONE,            // Disable logging
// };


//pub extern fn InitWindow(width: i32, height: i32, title:[*]const u8) void;

pub extern fn BeginDrawing() void;

pub extern fn ClearBackgroundExpanded(r: i32, g: i32, b: i32, a: i32) void;

//pub extern fn DrawTextExpanded(text:[]const u8, posX: i32, posY: i32, fontSize: i32, r: i32, g: i32, b: i32, a: i32) void;

pub extern fn SetTargetFPS(fps:i32) void;

pub extern fn EndDrawing() void;

pub extern fn CloseWindow() void;

//pub extern fn TraceLog(logLevel:i32, text:[*]const u8) void;

pub fn ClearBackground(color:Color) void {
ClearBackgroundExpanded(color.r, color.g, color.b, color.a);
}

// pub fn DrawText(text:[]const u8, posX:i32, posY:i32, fontSize:i32, color:Color) void {
//     //DrawTextExpanded(text, posX, posY, fontSize, color.r, color.g, color.b, color.a);
// }

File: examples/assemblyscript/src/raylib-wasm.ts

// Color, 4 components, R8G8B8A8 (32bit)
export class Color {
    r: u8
    g: u8
    b: u8
    a: u8

    constructor(r:u8, g:u8, b:u8, a:u8 = 255) {
        this.r = r
        this.g = g
        this.b = b
        this.a = a
    }
}

export const LIGHTGRAY: Color = new Color(200, 200, 200, 255)   // Light Gray
export const GRAY     : Color = new Color(130, 130, 130, 255)   // Gray
export const DARKGRAY : Color = new Color(80, 80, 80, 255)      // Dark Gray
export const YELLOW   : Color = new Color(253, 249, 0, 255)     // Yellow
export const GOLD     : Color = new Color(255, 203, 0, 255)     // Gold
export const ORANGE   : Color = new Color(255, 161, 0, 255)     // Orange
export const PINK     : Color = new Color(255, 109, 194, 255)   // Pink
export const RED      : Color = new Color(230, 41, 55, 255)     // Red
export const MAROON   : Color = new Color(190, 33, 55, 255)     // Maroon
export const GREEN    : Color = new Color(0, 228, 48, 255)      // Green
export const LIME     : Color = new Color(0, 158, 47, 255)      // Lime
export const DARKGREEN: Color = new Color(0, 117, 44, 255)      // Dark Green
export const SKYBLUE  : Color = new Color(102, 191, 255, 255)   // Sky Blue
export const BLUE     : Color = new Color(0, 121, 241, 255)     // Blue
export const DARKBLUE : Color = new Color(0, 82, 172, 255)      // Dark Blue
export const PURPLE   : Color = new Color(200, 122, 255, 255)   // Purple
export const VIOLET   : Color = new Color(135, 60, 190, 255)    // Violet
export const DARKPURPLE  : Color = new Color(112, 31, 126, 255)    // Dark Purple
export const BEIGE    : Color = new Color(211, 176, 131, 255)   // Beige
export const BROWN    : Color = new Color(127, 106, 79, 255)    // Brown
export const DARKBROWN: Color = new Color(76, 63, 47, 255)      // Dark Brown

export const WHITE    : Color = new Color(255, 255, 255, 255)   // White
export const BLACK    : Color = new Color(0, 0, 0, 255)         // Black
export const BLANK    : Color = new Color(0, 0, 0, 0)           // Blank (Transparent)
export const MAGENTA  : Color = new Color(255, 0, 255, 255)     // Magenta
export const RAYWHITE : Color = new Color(245, 245, 245, 255)   // My own White (raylib logo)

/*

WASM_EXPORT("Init")
void Init();

WASM_EXPORT("UpdateDrawFrame")
void UpdateDrawFrame();

WASM_EXPORT("Close")
void Close();

WASM_IMPORT("InitWindow")
void InitWindow(int width, int height, const char *title);

WASM_IMPORT("BeginDrawing")
void BeginDrawing();

WASM_IMPORT("ClearBackgroundExpanded")
void ClearBackgroundExpanded(uint8_t r, uint8_t g, uint8_t b, uint8_t a);

WASM_IMPORT("DrawTextExpanded")
void DrawTextExpanded(const char *text, int posX, int posY, int fontSize, uint8_t r, uint8_t g, uint8_t b, uint8_t a);

WASM_IMPORT("ClearBackground")
void ClearBackground();

WASM_IMPORT("SetTargetFPS")
void SetTargetFPS(int fps);

WASM_IMPORT("EndDrawing")
void EndDrawing();

WASM_IMPORT("CloseWindow")
void CloseWindow();

WASM_IMPORT("TraceLog")
void TraceLog(int logLevel, const char* text);

// Function wrappers
void ClearBackground(Color color);
void DrawText(const char* text, int posX, int posY, int fontSize, Color color);

#ifdef RAYLIB_WASM_IMPLEMENTATION
#ifndef RAYLIB_WASM_IMPLEMENTATION_ONCE
#define RAYLIB_WASM_IMPLEMENTATION_ONCE

void ClearBackground(Color color) {
    ClearBackgroundExpanded(color.r, color.g, color.b, color.a);
}

void DrawText(const char* text, int posX, int posY, int fontSize, Color color) {
    DrawTextExpanded(text, posX, posY, fontSize, color.r, color.g, color.b, color.a);
}

#endif  // RAYLIB_WASM_IMPLEMENTATION_ONCE
#endif  // RAYLIB_WASM_IMPLEMENTATION

#endif  // RAYLIB_WASM
*/

export const LOG_ALL = 0
export const LOG_TRACE = 1
export const LOG_DEBUG = 2
export const LOG_INFO = 3
export const LOG_WARNING = 4
export const LOG_ERROR = 5
export const LOG_FATAL = 6
export const LOG_NONE = 7

// since it takes a string arg, we convert to a UTF8 null-terminated buffer first
@external("env", "TraceLog")
declare function raylibTraceLog(logLevel: i32, text: ArrayBuffer) :void

export function TraceLog(logLevel: i32, text: string) :void {
    raylibTraceLog(logLevel, String.UTF8.encode(text, true))
}

